File: clojure/robot-name/_iterations/1/src/robot_name.clj
(ns robot-name)

(defprotocol ^:private Named
  (get-name [this])
  (set-name [this new-name]))

(deftype ^:private Robot [^:unsynchronized-mutable name]
  Named
  (get-name [this] name)
  (set-name [this new-name] (set! name new-name)))

(def ^:private robot-names (atom #{}))

(defn- random-char []
  (char (+ (int \A) (rand-int 26))))

(defn- random-digit []
  (char (+ (int \0) (rand-int 10))))

(defn- random-robot-name []
  (str (random-char) (random-char) (random-digit) (random-digit) (random-digit)))

(defn- unique-robot-name []
  (loop [name (random-robot-name)]
    (if (not (contains? @robot-names name))
      (do
        (swap! robot-names conj name)
        name)
      (recur (random-robot-name)))))

(defn robot []
  (Robot. (unique-robot-name)))

(defn robot-name [robot]
  (get-name robot))

(defn reset-name [robot]
  (set-name robot (unique-robot-name)))

File: julia/robot-name/_iterations/1/robot-name.jl
using Random: randstring

mutable struct Robot
    name::String
    Robot() = new(randomname())
end

function reset!(robot::Robot)
    robot.name = randomname()
    nothing
end

name(robot::Robot) = robot.name

const usednames = Set{String}()
const namespacelen = 26 * 26 * 10 * 10 * 10

function randomname()
    # Although unlikely, we might run out of unique names
    if length(usednames) == namespacelen
        error("Out of robot names")
    end

    while true
        name = randstring('A':'Z', 2) * randstring('0':'9', 3)
        if name ∉ usednames
            push!(usednames, name)
            return name
        end
    end
end
